### PR TITLE
Fallback on Zepto if jQuery is not available.

### PR DESCRIPTION
**Pull request by nickbaum**
Fallback on Zepto if jQuery is not available.

This is useful since popular frameworks like Foundation by default include Zepto rather than jQuery for browsers that support it.

Note that I have not tested form.js under Zepto. However, I believe this is still the right first step.

Zepto aims to be mostly compatible with jQuery, so there's a good chance that it will just work out of the box.

If not, we will discover specific Zepto compatibility issues, which may or may not be fixed, but either way the errors will be more helpful than "jQuery doesn't exist".

### CODE REVIEW

**Comment by malsup** [comment]
If you haven't tested this change with Zepto then what's the point?

**Comment by nickbaum** [comment]
Hi Mike,

You're absolutely right, that was rude of me. My thinking was that pushing this change would uncover any further incompatibilities with no downside for existing users, but I should have just tested it. Sorry about that, this is the first time I've actually submitted a pull request to an open source project.

I've now tested it, and in fact it did not work "out of the box" (so good call on rejecting it). The issue is line 208 where you select for non-empty inputs using [value!=""]. The attribute-not-equal is a jQuery specific selector, so it doesn't work in modern browsers' native querySelectorAll method.

However, with a small fix I was able to get it to work. As stated in the jQuery docs, [value!=""] is equivalent to not('[name="value"]'), which is compatible with CSS3 (and thus Zepto). In fact, using the 'not' syntax is slightly faster in modern browsers in jQuery as well.

http://api.jquery.com/attribute-not-equal-selector/

So to summarize, this pull request now has two one-line changes:
- import (window.jQuery || window.Zepto) rather than just jQuery
- change [value!=""] to not([value=""])

This time I have tested it, and it works correctly in my implementation. I can't guarantee that all the features work under Zepto since I don't use them all, but it's better than not working at all, and the fact that the jQuery docs state that the selector change is functionally equivalent gives me confidence that it won't break anything under jQuery. Let me know if you have a formal test suite and I will make sure that passes too.

Cheers,

-Nick

**Comment by nickbaum** [comment]
Hmm, I'm not seeing the second commit here yet. My understanding was that pushing it to the original branch would update the pull request, but this doesn't seem to have happened yet. Let me know if I need to do anything to update the pull request.

In the meantime, you can see the second commit here:

https://github.com/nickbaum/form/commit/a4aed089d99430691c5d83b9ea5872f5ff0f10be

-NIck

**Comment by nickbaum** [comment]
Ah, I believe the pull request should update if you re-open the pull request:

http://stackoverflow.com/questions/9790448/how-to-update-a-pull-request

Sorry for the confusion. I'm happy to send a fresh pull request if that is cleaner/easier.

-Nick

**Comment by nickbaum** [comment]
Hi Mike, I just wanted to check in on this. As mentioned, it's now a two line change and tested. Let me know if there's anything else I can do.

**Comment by malsup** [comment]
The problem is that my unit tests don't run under Zepto because I use a lot of jQuery-specific selectors, so I have a lot of work to do to re-write those tests.

**Comment by nickbaum** [comment]
Ah, that makes sense. Thanks for the explanation, Mike.

I don't see the unit tests in the public repo. If you're comfortable
pushing them, I'm happy to take a look. Otherwise I guess I'll give up on
this for now.

Just to summarize why supporting zepto be worthwhile, the
Foundationhttp://foundation.zurb.com/framework is Bootstrap's
main competitor http://responsive.vermilion.com/compare.php, used by
Mozilla, Pixar, NatGeo, etc. Because it defaults to Zepto for compatible
browsers, the form plugin doesn't work out of the box.

Either way, thanks for actively maintaining a great plugin.

-Nick

On Mon, Jul 29, 2013 at 2:37 PM, Mike Alsup notifications@github.comwrote:

> The problem is that my unit tests don't run under Zepto because I use a
> lot of jQuery-specific selectors, so I have a lot of work to do to re-write
> those tests.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/malsup/form/pull/325#issuecomment-21753710
> .

## 

How well do you know your parents?
Find out with www.storyworth.com
